restcomm/restcomm.interpreter/src/main/java/org/restcomm/connect/interpreter/VoiceInterpreter.java

```diff
@@ -443,6 +457,7 @@ private VoiceInterpreter(VoiceInterpreterParams params) {
         this.asImsUa = params.isAsImsUa();
         this.imsUaLogin = params.getImsUaLogin();
         this.imsUaPassword = params.getImsUaPassword();
+        this.msResponsePending = false;
```

Comment:
any reason why we need flags instead of proper state and transitions?. Adding boolean flags to model FSM states will diminish the FSM pattern, and lower maintainability

Reply:
I tried hard for a pure and dry FSM-based solution. But in the complexity of VI, that was the only functional solution i could came up with. Turns out that all FSM state classes manipulates shared resources that are highly sensitive, so introducing such mechanism was not possible without conflicts that break the call flow. Of course new ideas, alternatives and suggestions are always welcome! (although is important to keep in mind that the current state works!)

Reply:
ill have a look,maybe we need to add composite states to the FSM definition...

Reply:
@ghjansen I can see now that our FSM framework is quite limited (basically not an FSM).

It doesnt even declare how transitions are meant to be fired by FSM depending on incoming events, its just a collection of source and target states. The code needs to specify the transition to take by passing both the event and the target state, which is quite odd for a FSM pattern.

FSM doesnt support composite states. 

The code is also not separating different FSMs properly, so the common use of boolean flags to have different level of states and FSM data model.

Given this FSM status, i would certainly skip/defer any refactoring here, as it will require major effort beyond the scope of this PR.

We should certainly think about refactoring all of this with proper FSM, and potentially evaluate Squirrel as Henrique already did with RMS (all previous features supported...).

Reply:
Thank you very much @jaimecasero for all your considerations and guidance! Do you think that FSM/Squirrel refactoring would be a good topic to bring to the table in our next R&D offsite? I was thinking about that... not sure.

Reply:
i've been discussing FSMs with henrique, and specially in the context of RC. He wrote a good summary of RC issues with FSMs two years ago, that is mostly applicable today... 

The custom FSM "framework" we are using today is certainly not meeting the minimum requirements for a proper FSM pattern, just how it enforces you to check the current state, before telling the FSM to transition to target state,instead of just firing the event is painful. 

The great issue though is not the FSM pattern/lib itself, but how we have merged different FSMs into one, that will make any refactoring complex (regardles of which FSM lib we use). The good thing is that we have integration test to partly cover some of these actors behavior. Again, the FSM lib discussion is not key, but how we decouple the monolithic FSMs we have now, into smaller, independent,and cooperative FSMs...